Refactor errors, adding section for Message Function Errors (#774)
unicode-org/message-format-wg: spec/formatting.md

   resolve the value of the _expression_ as the result of that function call.

   If the call fails or does not return a valid value,

>    emit a _Message Function Error_.

Author: That seems a bit too opinionated? As in, an implementation could have a hierarchy of errors like Error > MessageFormatError > MessageFunctionError > BadOption, and have a need to use the MessageFunctionError class directly if user code throws a random error that's not a MessageFunctionError.

>    emit a _Message Function Error_.

Reviewer: ... in which case the "appropriate" error is the parent class of error. My suggestion permits the implementation to emit one of the subclasses of error and encourages the implementer to figure out which one it is. Generally speaking, if there's a random exception thrown, it'll be emitted as a FormattingError or SelectionError (indicating what the implementation was trying to do when it failed). This would even be helpful in debugging for functions that do both selection and formatting, since it would tell me which part of the function the error took place in.

   1. Else:
      1. Let `nomatch` be a resolved value for which selection always fails.
      1. Append `nomatch` as the last element of the list `res`.

>       1. Emit a _Message Function Error_.

Author: ```suggestion
      1. Emit a _Selection Error_.
```

Resolved values cannot always be formatted by a given implementation.
When such an error occurs during _formatting_,

> an implementation SHOULD emit a _Message Function Error_ and produce a

Author: We'll be able to test for the more general error class as well. @mradbourne or I will file a follow-up PR to his current one to allow for that.

> an implementation SHOULD emit a _Message Function Error_ and produce a

Reviewer: Agreed that we should create a test for it. My comment, though, is that we should nail this one down somewhat. Or do you disagree?

> an implementation SHOULD emit a _Message Function Error_ and produce a

Author: I disagree. I think leaving it up to the function implementer to pick between the options will provide the best user experience, and it allows for lazy implementations to e.g. only check operands and options until they're absolutely required, and to emit Bad Operand or Bad Option errors at any time.

> Resolved values cannot always be formatted by a given implementation.
> When such an error occurs during _formatting_,
> an implementation SHOULD emit a _Message Function Error_ and produce a

Reviewer: This is a very specific bit of guidance: if there is a fully resolved value _and_ the function determines it cannot be formatted by that function during the formatting process, isn't that the definition of a Bad Operand? Plus, this is `SHOULD` not `MUST`, so an implementation is still permitted to emit something else if it has good reason. 

If I'm an implementer and I'm reading the spec, I'm going to be looking for `MUSTard` that I am required to implement (and the tests that go with it). If we intend to be wiggly, we need to make it clear. In that case, it would look more like:

```suggestion
Resolved values cannot always be formatted by a given implementation.
When a _function_ determines that it does not support the resolve value of its _operand_
its implementation SHOULD emit an informative error, such as _Bad Operand_,
and produce a
```

> an implementation SHOULD emit a _Message Function Error_ and produce a

Author: ```suggestion
an implementation SHOULD emit an apropriate _Message Function Error_ and produce a
```

> an implementation SHOULD emit a _Message Function Error_ and produce a

Author: I've updated the text here to now say:
> When such an error occurs during _formatting_,
> an implementation SHOULD emit an appropriate _Message Function Error_ and produce ...

which is in line with @aphillips's suggestion in https://github.com/unicode-org/message-format-wg/pull/774#discussion_r1576395507.
